Add DiagnosticLevel used to filter how verbose toStringDeep output is.

jacob314 opened the pull request:
The 'creator' property is set with priority DiagnosticLevel.debug instead of DiagnosticLevel.info so that it can be filtered out in the future for output shown by the Inspector.
Properties that match their default value now have DiagnosticLevel.fine instead of being "hidden"
so you now have the ability to see all properties that match their default value which can sometimes be useful particularly if you are only viewing properties for a single object instead of a full tree.

Only review the very last patch in this diff. The previous patches are patches from the refactoring of Diagnostics for the semantics tree which this CL has to be layered on top of to avoid duplicated tweaks.

jacob314 committed:
Add DiagnosticLevel used to filter how verbose toStringDeep output for
Diagnostics object is.

Hixie commented on packages/flutter/lib/rendering.dart:
```diff
@@ -27,6 +27,7 @@ export 'package:flutter/foundation.dart' show
   ValueChanged,
   ValueGetter,
   ValueSetter;
+export 'package:flutter/foundation.dart' show DiagnosticLevel;
```
why not add this to the list above?

jacob314 replied:
done. In other files I was seeing repeated export statements for blocks of names that weren't conceptually related. I don't have a preference either way.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -6,6 +6,41 @@ import 'package:meta/meta.dart';
+///
+/// Trees of Flutter diagnostics can be very large so filtering the Diagnostics
+/// shown matters. Typically filtering to only show diagnostics with at least
+/// level debug is appropriate.
```
`[debug]`

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -6,6 +6,41 @@ import 'package:meta/meta.dart';
+/// The various priority levels used to filter which diagnostics are shown and
+/// omitted.
+///
+/// Trees of Flutter diagnostics can be very large so filtering the Diagnostics
```
if Diagnostics is a reference to an object, link to it; otherwise, lowercase d.

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -6,6 +6,41 @@ import 'package:meta/meta.dart';
+/// shown matters. Typically filtering to only show diagnostics with at least
+/// level debug is appropriate.
+enum DiagnosticLevel {
+  /// Use this level for diagnostic properties that match their default value
```
Add a paragraph above the current paragraph for `fine` that describes this value, e.g. "Trivial or redundant diagnostics information."

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -577,14 +612,18 @@ const _NoDefaultValue kNoDefaultValue = const _NoDefaultValue();
   /// Initializes the object.
   ///
-  /// The [style], [showName] and [showSeparator] arguments must not be null.
+  /// The [style], [showName], [showSeparator] and [level] arguments must not
```
comma before and

jacob314 replied:
done. Added oxford comma here and elsewhere in this file.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -634,10 +676,20 @@ abstract class DiagnosticsNode {
 
+  final DiagnosticLevel _defaultLevel;
+
+  DiagnosticLevel get level => _defaultLevel;
```
docs

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -634,10 +676,20 @@ abstract class DiagnosticsNode {
+  }
+
   /// Whether the diagnostics should be hidden when showing the default
   /// view of a tree.
```
this should talk about its relationship with level

jacob314 replied:
I realized this method would be cleaner to document if I removed the `hidden` property from DiagnosticNode. Added [DiagnosticLevel.hidden] to serve the same role. This makes filtering more standard. I also cleaned up toString output for FlagProperty and ObjectFlagProperty cases where the output previously didn't matter as the properties were hidden.

One open question is whether I should make the breaking change to remove the hidden constructor argument as well. It is redundant but is also convenient shorthand for toggling whether a property should have DiagnosticLevel.hidden.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -920,17 +998,19 @@ class MessageProperty extends DiagnosticsProperty<Null> {
   /// Create a diagnostics property for strings.
   ///
-  /// The [showName], [hidden], and [quoted] arguments must not be null.
+  /// The [showName], [hidden], [quoted] and [level] arguments must not be null.
```
comma before and

jacob314 replied:
done here and elsewhere.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -6,6 +6,41 @@ import 'package:meta/meta.dart';
+  /// Very important diagnostics that indicate problematic property values.
+  ///
+  /// For example, use if you would write the property description
+  /// message in ALL CAPS.
```
there are some examples of this in object.dart and box.dart, should they be updated too?

jacob314 replied:
updated the example in box.dart
the object.dart example doesn't quite apply as NEEDS-LAYOUT, NEEDS-PAINT and DETATCHED
aren't currently possible to handle as properties as they are in the toStringShort instead of the regular body of properties.We could be clever and display warning level FlagProperties first in the toString output if we wanted.

Hixie commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -2855,18 +2863,21 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   void debugFillProperties(DiagnosticPropertiesBuilder description) {
-    description.add(new DiagnosticsProperty<dynamic>('creator', debugCreator, defaultValue: null));
+    description.add(new DiagnosticsProperty<dynamic>('creator', debugCreator, defaultValue: null, level: DiagnosticLevel.debug));
     description.add(new DiagnosticsProperty<ParentData>('parentData', parentData, tooltip: _debugCanParentUseSize == true ? "can use size" : null));
```
`s/"/'/` on this line

jacob314 replied:
done

jacob314 commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -6,6 +6,41 @@ import 'package:meta/meta.dart';
+enum DiagnosticLevel {
+  /// Use this level for diagnostic properties that match their default value
+  /// and other cases where showing a diagnostic would not add value.
+  fine,
```
done

jacob314 commented:
PTAL

jacob314 commented:
Resolved merge conflicts. Added warnings for additional 'MISSING' properties added with the merge. Added missingIfNull bool named parameter to DiagnosticsNode that handles this pattern.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -6,6 +6,55 @@ import 'package:meta/meta.dart';
+  ///
+  /// Use this level for diagnostic properties that match their default value
+  /// and other cases where showing a diagnostic would not add value.
+  fine,
```
should we merge hidden and fine? their descriptions sound very similar.

jacob314 replied:
I think the problem was with the quality of my description of the three levels. I've written more to clarify the difference between hidden, fine, and debug.
TL:DR;
hidden: property display may be ugly or the value may simply be misleading (e.g. some other property takes priority so the property has no impact.
fine: property display is great but the value is known to be uninteresting.
debug: property value is interesting but it doesn't meet the bar to be a higher priority level likely due to being very verbose.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -6,6 +6,55 @@ import 'package:meta/meta.dart';
+  /// and other cases where showing a diagnostic would not add value.
+  fine,
+
+  /// Diagnostics that should only be shown when performing fine grained
```
the use of the word "fine" in the definition of "debug" is a bit suggestive too.

jacob314 replied:
sorry about the bad comments. Updated them to convey the actual distinction I'm getting at. Generally, if you set the minLevel to fine, you should still see beautiful diagnostics just a lot more info and some diagnostics that maybe if you were an experienced user you wouldn't be interested in as they are the default value.
If you set minLevel to hidden you see a bunch more kinda ugly looking and barely functional diagnostics, but you are displaying diagnostics that the component authors explicitly decided to hide so it is your own fault.
If you set minLevel to debug you have everything you need to debug assuming you are savvy enough to understand what the defaultValues are, etc.
If you set minLevel to info you should have enough info to browse the structure.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -596,13 +646,19 @@ abstract class DiagnosticsNode {
+  /// At most one of `hidden` and [level] should be specified. Setting `hidden`
+  /// to `true` is equivalent to setting [level] to [DiagnosticLevel.hidden].
+  /// If neither `hidden` or [level] is set, [level] defaults to
+  /// [DiagnosticsLevel.info].
```
am i missing something? I don't see a `hidden` on this constructor.

jacob314 replied:
Removed all references to `hidden`

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -1406,36 +1618,42 @@ typedef T ComputePropertyValueCallback<T>();
     String name,
     T value, {
     String description,
-    bool hidden: false,
```
you've already broken back-compat by removing the "hidden" property getter, I'd just go all the way and remove the argument too. I don't think it provides much value.

jacob314 replied:
Removed the hidden argument. Most of the cases where it was used were IterableProperty cases better handled by `ifEmpty: null` which I've added support for.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -1543,6 +1769,9 @@ class DiagnosticsProperty<T> extends DiagnosticsNode {
   /// generating the string description.
   final String tooltip;
 
+  /// Whether a value of null should trigger a warning.
```
these docs talk should talk about the implications on `level`

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -1600,29 +1829,31 @@ class DiagnosticsProperty<T> extends DiagnosticsNode {
-  /// [hidden] as the [value] is uninteresting.
+  /// If the [value] of the property equals [defaultValue] the priority [level]
+  /// of the property is downgraded to [DiagnosticLevel.fine] as the property
+  /// value is uninteresting.
```
these docs should talk about `missingIfNull`

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -1448,30 +1666,38 @@ class DiagnosticsProperty<T> extends DiagnosticsNode {
+  /// to `true` is equivalent to setting [level] to [DiagnosticLevel.hidden].
+  /// If neither `hidden` or [level] is set, [level] defaults to
+  /// [DiagnosticsLevel.info].
   DiagnosticsProperty.lazy(
```
these docs should talk about how the "level" argument is just a suggestion and how it can be overridden by other things.

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -634,9 +691,20 @@ abstract class DiagnosticsNode {
+  /// Typically this only makes sense to set to a different value than
+  /// [DiagnosticLevel.info] for diagnostics representing properties where some
+  /// properties may be only marginally relevant.
+  DiagnosticLevel get level => DiagnosticLevel.info;
```
either:
* this should probably mention how some subclasses have a "level" argument to their constructor which influences the value returned here, but that other factors also influence it,
or:
* the subclasses should provide their own overriding docs for this member when they change its behaviour

jacob314 replied:
Good idea. Done.

Hixie commented:
I recommend removing the "hidden" arguments.
LGTM, modulo that and the suggested doc fixes.
